File: android/src/main/java/expo/modules/llmmediapipe/ExpoLlmMediapipeModule.kt
package expo.modules.llmmediapipe

import android.content.Context
import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.*
import org.json.JSONObject
import java.io.BufferedInputStream

private const val TAG = "ExpoLlmMediapipe"
private const val DOWNLOAD_DIRECTORY = "llm_models"

class ExpoLlmMediapipeModule : Module() {
  private var nextHandle = 1
  private val modelMap = mutableMapOf<Int, LlmInferenceModel>()

  // Define these functions at class level, not in the definition block
  private fun createInferenceListener(modelHandle: Int): InferenceListener {
    return object : InferenceListener {
      override fun logging(model: LlmInferenceModel, message: String) {
        sendEvent("logging", mapOf(
          "handle" to modelHandle,
          "message" to message
        ))
      }
      
      override fun onError(model: LlmInferenceModel, requestId: Int, error: String) {
        sendEvent("onErrorResponse", mapOf(
          "handle" to modelHandle,
          "requestId" to requestId,
          "error" to error
        ))
      }
      
      override fun onResults(model: LlmInferenceModel, requestId: Int, response: String) {
        sendEvent("onPartialResponse", mapOf(
          "handle" to modelHandle,
          "requestId" to requestId,
          "response" to response
        ))
      }
    }
  }
  
  private fun copyFileToInternalStorageIfNeeded(modelName: String, context: Context): File {
    val outputFile = File(context.filesDir, modelName)

    // Check if the file already exists
    if (outputFile.exists()) {
      // The file already exists, no need to copy again
      sendEvent("logging", mapOf(
        "message" to "File already exists: ${outputFile.path}, size: ${outputFile.length()}"
      ))
      return outputFile
    }

    try {
      val assetList = context.assets.list("") ?: arrayOf()
      sendEvent("logging", mapOf(
        "message" to "Available assets: ${assetList.joinToString()}"
      ))
      
      if (!assetList.contains(modelName)) {
        val errorMsg = "Asset file $modelName does not exist in assets"
        sendEvent("logging", mapOf("message" to errorMsg))
        throw IllegalArgumentException(errorMsg)
      }

      sendEvent("logging", mapOf(
        "message" to "Copying asset $modelName to ${outputFile.path}"
      ))
      
      // File doesn't exist, proceed with copying
      context.assets.open(modelName).use { inputStream ->
        FileOutputStream(outputFile).use { outputStream -> 
          val buffer = ByteArray(1024)
          var read: Int
          var total = 0
          
          while (inputStream.read(buffer).also { read = it } != -1) {
            outputStream.write(buffer, 0, read)
            total += read
            
            if (total % (1024 * 1024) == 0) { // Log every MB
              sendEvent("logging", mapOf(
                "message" to "Copied $total bytes so far"
              ))
            }
          }
          
          sendEvent("logging", mapOf(
            "message" to "Copied $total bytes total"
          ))
        }
      }
    } catch (e: Exception) {
      sendEvent("logging", mapOf(
        "message" to "Error copying file: ${e.message}"
      ))
      throw e
    }

    return outputFile
  }

  // Model directory management
  private fun getModelDirectory(): File {
    val modelDir = File(appContext.reactContext!!.filesDir, DOWNLOAD_DIRECTORY)
    if (!modelDir.exists()) {
      modelDir.mkdirs()
    }
    return modelDir
  }
  
  private fun getModelFile(modelName: String): File {
    return File(getModelDirectory(), modelName)
  }
  
  // Create model internal helper method
  private fun createModelInternal(modelPath: String, maxTokens: Int, topK: Int, temperature: Double, randomSeed: Int): Int {
    val modelHandle = nextHandle++
    val model = LlmInferenceModel(
      appContext.reactContext!!,
      modelPath,
      maxTokens,
      topK,
      temperature.toFloat(),
      randomSeed,
      inferenceListener = createInferenceListener(modelHandle)
    )
    modelMap[modelHandle] = model
    return modelHandle
  }

  // Track active downloads
  private val activeDownloads = mutableMapOf<String, Job>()

  override fun definition() = ModuleDefinition {
    Name("ExpoLlmMediapipe")

    Constants(
      "PI" to Math.PI
    )

    Events("onChange", "onPartialResponse", "onErrorResponse", "logging", "downloadProgress")

    Function("hello") {
      "Hello world from MediaPipe LLM! 👋"
    }

    AsyncFunction("createModel") { modelPath: String, maxTokens: Int, topK: Int, temperature: Double, randomSeed: Int, promise: Promise ->
      try {
        val modelHandle = nextHandle++
        
        // Log that we're creating a model
        sendEvent("logging", mapOf(
          "handle" to modelHandle,
          "message" to "Creating model from path: $modelPath"
        ))
        
        val model = LlmInferenceModel(
          appContext.reactContext!!,
          modelPath,
          maxTokens,
          topK,
          temperature.toFloat(),
          randomSeed,
          inferenceListener = createInferenceListener(modelHandle)
        )
        modelMap[modelHandle] = model
        promise.resolve(modelHandle)
      } catch (e: Exception) {
        // Log the error
        sendEvent("logging", mapOf(
          "message" to "Model creation failed: ${e.message}"
        ))
        promise.reject("MODEL_CREATION_FAILED", e.message ?: "Unknown error", e)
      }
    }

    AsyncFunction("createModelFromAsset") { modelName: String, maxTokens: Int, topK: Int, temperature: Double, randomSeed: Int, promise: Promise ->
      try {
        // Log that we're creating a model from asset
        sendEvent("logging", mapOf(
          "message" to "Creating model from asset: $modelName"
        ))
        
        val modelPath = copyFileToInternalStorageIfNeeded(modelName, appContext.reactContext!!).path
        
        sendEvent("logging", mapOf(
          "message" to "Model file copied to: $modelPath"
        ))
        
        val modelHandle = nextHandle++
        val model = LlmInferenceModel(
          appContext.reactContext!!,
          modelPath,
          maxTokens,
          topK,
          temperature.toFloat(),
          randomSeed,
          inferenceListener = createInferenceListener(modelHandle)
        )
        modelMap[modelHandle] = model
        promise.resolve(modelHandle)
      } catch (e: Exception) {
        // Log the error
        sendEvent("logging", mapOf(
          "message" to "Model creation from asset failed: ${e.message}"
        ))
        promise.reject("MODEL_CREATION_FAILED", e.message ?: "Unknown error", e)
      }
    }

    AsyncFunction("releaseModel") { handle: Int, promise: Promise ->
      try {
        val removed = modelMap.remove(handle) != null
        if (removed) {
          promise.resolve(true)
        } else {
          promise.reject("INVALID_HANDLE", "No model found for handle $handle", null)
        }
      } catch (e: Exception) {
        promise.reject("RELEASE_FAILED", e.message ?: "Unknown error", e)
      }
    }

    AsyncFunction("generateResponse") { handle: Int, requestId: Int, prompt: String, promise: Promise ->
      try {
        val model = modelMap[handle]
        if (model == null) {
          promise.reject("INVALID_HANDLE", "No model found for handle $handle", null)
          return@AsyncFunction
        }
        
        sendEvent("logging", mapOf(
          "handle" to handle,
          "message" to "Generating response with prompt: ${prompt.take(30)}..."
        ))
        
        // Use the synchronous version
        val response = model.generateResponse(requestId, prompt)
        promise.resolve(response)
      } catch (e: Exception) {
        sendEvent("logging", mapOf(
          "handle" to handle,
          "message" to "Generation error: ${e.message}"
        ))
        promise.reject("GENERATION_FAILED", e.message ?: "Unknown error", e)
      }
    }

    AsyncFunction("generateResponseAsync") { handle: Int, requestId: Int, prompt: String, promise: Promise ->
      try {
        val model = modelMap[handle]
        if (model == null) {
          promise.reject("INVALID_HANDLE", "No model found for handle $handle", null)
          return@AsyncFunction
        }
        
        sendEvent("logging", mapOf(
          "handle" to handle,
          "requestId" to requestId,
          "message" to "Starting async generation with prompt: ${prompt.take(30)}..."
        ))
        
        // Use the async version with callback and event emission
        try {
          model.generateResponseAsync(requestId, prompt) { result ->
            try {
              if (result.isEmpty()) {
                sendEvent("logging", mapOf(
                  "handle" to handle,
                  "requestId" to requestId,
                  "message" to "Generation completed but returned empty result"
                ))
                promise.reject("GENERATION_FAILED", "Failed to generate response", null)
              } else {
                sendEvent("logging", mapOf(
                  "handle" to handle,
                  "requestId" to requestId,
                  "message" to "Generation completed successfully with ${result.length} characters"
                ))
                
                // We don't resolve with the final result here anymore
                // The client will assemble the full response from streaming events
                promise.resolve(true)  // Just send success signal
              }
            } catch (e: Exception) {
              sendEvent("logging", mapOf(
                "handle" to handle,
                "requestId" to requestId,
                "message" to "Error in async result callback: ${e.message}"
              ))
              // Only reject if not already settled
              promise.reject("GENERATION_ERROR", e.message ?: "Unknown error", e)
            }
          }
        } catch (e: Exception) {
          sendEvent("logging", mapOf(
            "handle" to handle,
            "requestId" to requestId,
            "message" to "Exception during generateResponseAsync call: ${e.message}"
          ))
          promise.reject("GENERATION_ERROR", e.message ?: "Unknown error", e)
        }
      } catch (e: Exception) {
        sendEvent("logging", mapOf(
          "handle" to handle,
          "message" to "Outer exception in generateResponseAsync: ${e.message}"
        ))
        promise.reject("GENERATION_ERROR", e.message ?: "Unknown error", e)
      }
    }

    // Check if model is downloaded
    AsyncFunction("isModelDownloaded") { modelName: String, promise: Promise ->
      val modelFile = getModelFile(modelName)
      promise.resolve(modelFile.exists() && modelFile.length() > 0)
    }
    
    // Get list of downloaded models
    AsyncFunction("getDownloadedModels") { promise: Promise ->
      val models = getModelDirectory().listFiles()?.map { it.name } ?: emptyList()
      promise.resolve(models)
    }
    
    // Delete downloaded model
    AsyncFunction("deleteDownloadedModel") { modelName: String, promise: Promise ->
      val modelFile = getModelFile(modelName)
      val result = if (modelFile.exists()) modelFile.delete() else false
      promise.resolve(result)
    }
    
    // Download model from URL
    AsyncFunction("downloadModel") { url: String, modelName: String, options: Map<String, Any>?, promise: Promise ->
      val modelFile = getModelFile(modelName)
      val overwrite = (options?.get("overwrite") as? Boolean) ?: false
      
      // Check if already downloading
      if (activeDownloads.containsKey(modelName)) {
        promise.reject("ERR_ALREADY_DOWNLOADING", "This model is already being downloaded", null)
        return@AsyncFunction
      }
      
      // Check if already exists
      if (modelFile.exists() && !overwrite) {
        promise.resolve(true)
        return@AsyncFunction
      }
      
      // Start download in coroutine
      val downloadJob = CoroutineScope(Dispatchers.IO).launch {
        try {
          val connection = URL(url).openConnection() as HttpURLConnection
          
          // Add custom headers if provided
          (options?.get("headers") as? Map<String, Any>)?.let { headers ->
            headers.forEach { (key, value) ->
              connection.setRequestProperty(key, value.toString())
            }
          }
          
          connection.connectTimeout = (options?.get("timeout") as? Number)?.toInt() ?: 30000
          connection.connect()
          
          val contentLength = connection.contentLength.toLong()
          val input = BufferedInputStream(connection.inputStream)
          val tempFile = File(modelFile.absolutePath + ".temp")
          val output = FileOutputStream(tempFile)
          
          val buffer = ByteArray(8192)
          var total: Long = 0
          var count: Int
          var lastUpdateTime = System.currentTimeMillis()
          
          while (input.read(buffer).also { count = it } != -1) {
            if (isActive.not()) {
              // Download was cancelled
              output.close()
              input.close()
              tempFile.delete()
              sendEvent("downloadProgress", mapOf(
                "modelName" to modelName,
                "url" to url,
                "status" to "cancelled"
              ))
              return@launch
            }
            
            total += count
            output.write(buffer, 0, count)
            
            // Send progress updates, throttled to avoid too many events
            val currentTime = System.currentTimeMillis()
            if (currentTime - lastUpdateTime > 100) { // Every 100ms
              lastUpdateTime = currentTime
              val progress = if (contentLength > 0) total.toDouble() / contentLength.toDouble() else 0.0
              sendEvent("downloadProgress", mapOf(
                "modelName" to modelName,
                "url" to url,
                "bytesDownloaded" to total,
                "totalBytes" to contentLength,
                "progress" to progress,
                "status" to "downloading"
              ))
            }
          }
          
          // Close streams
          output.flush()
          output.close()
          input.close()
          
          // Rename temp file to final file
          if (modelFile.exists()) {
            modelFile.delete()
          }
          tempFile.renameTo(modelFile)
          
          // Notify completion
          sendEvent("downloadProgress", mapOf(
            "modelName" to modelName,
            "url" to url,
            "bytesDownloaded" to modelFile.length(),
            "totalBytes" to modelFile.length(),
            "progress" to 1.0,
            "status" to "completed"
          ))
          
          withContext(Dispatchers.Main) {
            promise.resolve(true)
          }
        } catch (e: Exception) {
          Log.e(TAG, "Error downloading model: ${e.message}", e)
          sendEvent("downloadProgress", mapOf(
            "modelName" to modelName,
            "url" to url,
            "status" to "error",
            "error" to (e.message ?: "Unknown error")
          ))
          withContext(Dispatchers.Main) {
            promise.reject("ERR_DOWNLOAD", "Failed to download model: ${e.message}", e)
          }
        } finally {
          activeDownloads.remove(modelName)
        }
      }
      
      activeDownloads[modelName] = downloadJob
    }
    
    // Cancel download
    AsyncFunction("cancelDownload") { modelName: String, promise: Promise ->
      val job = activeDownloads[modelName]
      if (job != null) {
        job.cancel()
        activeDownloads.remove(modelName)
        promise.resolve(true)
      } else {
        promise.resolve(false)
      }
    }
    
    // Create model from downloaded file
    AsyncFunction("createModelFromDownloaded") { modelName: String, maxTokens: Int?, topK: Int?, temperature: Double?, randomSeed: Int?, promise: Promise ->
      val modelFile = getModelFile(modelName)
      
      if (!modelFile.exists()) {
        promise.reject("ERR_MODEL_NOT_FOUND", "Model $modelName is not downloaded", null)
        return@AsyncFunction
      }
      
      try {
        val handle = createModelInternal(
          modelFile.absolutePath,
          maxTokens ?: 1024,
          topK ?: 40,
          temperature ?: 0.7,
          randomSeed ?: 42
        )
        // Explicitly cast to avoid ambiguity
        promise.resolve(handle as Int)
      } catch (e: Exception) {
        Log.e(TAG, "Error creating model from downloaded file: ${e.message}", e)
        promise.reject("ERR_CREATE_MODEL", "Failed to create model: ${e.message}", e)
      }
    }
  }
}

File: ios/ExpoLlmMediapipeModule.swift
import ExpoModulesCore
import MediaPipeTasksGenAI

public class ExpoLlmMediapipeModule: Module {
  private var modelMap = [Int: LlmInferenceModel]()
  private var nextHandle = 1
  private var activeDownloads: [String: URLSessionDownloadTask] = [:]
  private var downloadObservers: [String: NSKeyValueObservation] = [:]

  private func getModelDirectory() -> URL {
    let fileManager = FileManager.default
    let documentsURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
    let modelDirURL = documentsURL.appendingPathComponent("llm_models")
    if !fileManager.fileExists(atPath: modelDirURL.path) {
      try? fileManager.createDirectory(at: modelDirURL, withIntermediateDirectories: true)
    }
    return modelDirURL
  }

  private func getModelURL(modelName: String) -> URL {
    return getModelDirectory().appendingPathComponent(modelName)
  }

  public func definition() -> ModuleDefinition {
    Name("ExpoLlmMediapipe")

    Events("downloadProgress", "onChange", "onPartialResponse", "onErrorResponse", "logging")

    Function("hello") {
      return "Hello world from MediaPipe LLM on iOS! 👋"
    }

    AsyncFunction("createModel", createModel)
    AsyncFunction("createModelFromAsset", createModelFromAsset)
    AsyncFunction("releaseModel", releaseModel)
    AsyncFunction("generateResponse", generateResponse)
    AsyncFunction("generateResponseAsync", generateResponseAsync)
    AsyncFunction("isModelDownloaded", isModelDownloaded)
    AsyncFunction("getDownloadedModels", getDownloadedModels)
    AsyncFunction("deleteDownloadedModel", deleteDownloadedModel)
    AsyncFunction("downloadModel", downloadModel)
    AsyncFunction("cancelDownload", cancelDownload)
    AsyncFunction("createModelFromDownloaded", createModelFromDownloaded)

    Function("printModelInfo") {
      let modelPaths = try? FileManager.default.contentsOfDirectory(atPath: Bundle.main.resourcePath ?? "")
      let models = modelPaths?.filter { $0.hasSuffix(".bin") || $0.hasSuffix(".task") }
      return [
        "bundle_path": Bundle.main.resourcePath ?? "unknown",
        "models_found": models ?? [],
        "api_available": NSClassFromString("LlmInference") != nil
      ]
    }
  }

  private func createModel(modelPath: String, maxTokens: Int, topK: Int, temperature: Double, randomSeed: Int, promise: Promise) {
    do {
      let modelHandle = nextHandle
      nextHandle += 1
      let model = try LlmInferenceModel(
        modelPath: modelPath,
        maxTokens: maxTokens,
        topK: topK,
        temperature: Float(temperature),
        randomSeed: randomSeed,
        eventEmitter: { [weak self] eventName, params in self?.sendEvent(eventName, params) },
        modelHandle: modelHandle
      )
      modelMap[modelHandle] = model
      promise.resolve(modelHandle)
    } catch {
      promise.reject("MODEL_ERROR", "Failed to create model: \(error)")
    }
  }

  private func createModelFromAsset(modelName: String, maxTokens: Int, topK: Int, temperature: Double, randomSeed: Int, promise: Promise) {
    guard let modelURL = Bundle.main.url(forResource: modelName, withExtension: nil) else {
      promise.reject("MODEL_NOT_FOUND", "Model not found in app bundle: \(modelName)")
      return
    }
    createModel(modelPath: modelURL.path, maxTokens: maxTokens, topK: topK, temperature: temperature, randomSeed: randomSeed, promise: promise)
  }

  private func releaseModel(handle: Int, promise: Promise) {
    if modelMap[handle] != nil {
      modelMap.removeValue(forKey: handle)
      promise.resolve(true)
    } else {
      promise.reject("INVALID_HANDLE", "No model found for handle \(handle)")
    }
  }

  private func generateResponse(handle: Int, requestId: Int, prompt: String, promise: Promise) {
    guard let model = modelMap[handle] else {
      promise.reject("INVALID_HANDLE", "No model found for handle \(handle)")
      return
    }
    do {
      try model.generateResponse(requestId: requestId, prompt: prompt) { result in
        switch result {
        case .success(let response):
          promise.resolve(response)
        case .failure(let error):
          promise.reject("GENERATION_ERROR", error.localizedDescription)
        }
      }
    } catch {
      promise.reject("GENERATION_ERROR", "Failed to generate response: \(error.localizedDescription)")
    }
  }

  private func generateResponseAsync(handle: Int, requestId: Int, prompt: String, promise: Promise) {
    guard let model = modelMap[handle] else {
      promise.reject("INVALID_HANDLE", "No model found for handle \(handle)")
      return
    }
    do {
      try model.generateStreamingResponse(requestId: requestId, prompt: prompt) { completed in
        if completed {
          promise.resolve(true)
        } else {
          promise.reject("GENERATION_INCOMPLETE", "Generation did not complete successfully")
        }
      }
    } catch {
      promise.reject("GENERATION_ERROR", "Failed to generate response: \(error.localizedDescription)")
    }
  }

  private func isModelDownloaded(modelName: String, promise: Promise) {
    let modelURL = getModelURL(modelName: modelName)
    let exists = FileManager.default.fileExists(atPath: modelURL.path)
    promise.resolve(exists)
  }

  private func getDownloadedModels(promise: Promise) {
    let modelDir = getModelDirectory()
    let fileManager = FileManager.default
    do {
      let fileURLs = try fileManager.contentsOfDirectory(at: modelDir, includingPropertiesForKeys: nil)
      let modelNames = fileURLs.map { $0.lastPathComponent }
      promise.resolve(modelNames)
    } catch {
      promise.resolve([])
    }
  }

  private func deleteDownloadedModel(modelName: String, promise: Promise) {
    let modelURL = getModelURL(modelName: modelName)
    do {
      try FileManager.default.removeItem(at: modelURL)
      promise.resolve(true)
    } catch {
      promise.resolve(false)
    }
  }

  private func downloadModel(url: String, modelName: String, options: [String: Any]?, promise: Promise) {
      let modelURL = self.getModelURL(modelName: modelName)
      let overwrite = options?["overwrite"] as? Bool ?? false
      
      // Check if already downloading
      if self.activeDownloads[modelName] != nil {
        promise.reject("ERR_ALREADY_DOWNLOADING", "This model is already being downloaded")
        return
      }
      
      // Check if already exists
      if FileManager.default.fileExists(atPath: modelURL.path) && !overwrite {
        promise.resolve(true)
        return
      }
      
      // Create URL session for download
      guard let downloadURL = URL(string: url) else {
        promise.reject("ERR_INVALID_URL", "Invalid URL provided")
        return
      }
      
      var request = URLRequest(url: downloadURL)
      
      // Add custom headers if provided
      if let headers = options?["headers"] as? [String: String] {
        for (key, value) in headers {
          request.setValue(value, forHTTPHeaderField: key)
        }
      }
      
      // Set timeout
      if let timeout = options?["timeout"] as? TimeInterval {
        request.timeoutInterval = timeout / 1000.0
      }
      
      // Create download task
      let session = URLSession(configuration: .default)
      let task = session.downloadTask(with: request) { (tempURL, response, error) in
        // Remove from active downloads
        self.activeDownloads.removeValue(forKey: modelName)
        self.downloadObservers.removeValue(forKey: modelName)?.invalidate()
        
        if let error = error {
          self.sendEvent("downloadProgress", [
            "modelName": modelName,
            "url": url,
            "status": "error",
            "error": error.localizedDescription
          ])
          promise.reject("ERR_DOWNLOAD", "Download failed: \(error.localizedDescription)")
          return
        }
        
        guard let tempURL = tempURL else {
          self.sendEvent("downloadProgress", [
            "modelName": modelName,
            "url": url,
            "status": "error",
            "error": "No file downloaded"
          ])
          promise.reject("ERR_DOWNLOAD", "Download failed: No file received")
          return
        }
        
        do {
          // If file already exists, remove it
          if FileManager.default.fileExists(atPath: modelURL.path) {
            try FileManager.default.removeItem(at: modelURL)
          }
          
          // Move downloaded file to final location
          try FileManager.default.moveItem(at: tempURL, to: modelURL)
          
          self.sendEvent("downloadProgress", [
            "modelName": modelName,
            "url": url,
            "bytesDownloaded": try FileManager.default.attributesOfItem(atPath: modelURL.path)[.size] as? Int64 ?? 0,
            "totalBytes": try FileManager.default.attributesOfItem(atPath: modelURL.path)[.size] as? Int64 ?? 0,
            "progress": 1.0,
            "status": "completed"
          ])
          
          promise.resolve(true)
        } catch {
          self.sendEvent("downloadProgress", [
            "modelName": modelName,
            "url": url,
            "status": "error",
            "error": error.localizedDescription
          ])
          promise.reject("ERR_SAVE_FILE", "Failed to save downloaded model: \(error.localizedDescription)")
        }
      }
      
      // Observe download progress
      let observation = task.progress.observe(\.fractionCompleted) { progress, _ in
        DispatchQueue.main.async {
          self.sendEvent("downloadProgress", [
            "modelName": modelName,
            "url": url,
            "bytesDownloaded": task.countOfBytesReceived,
            "totalBytes": task.countOfBytesExpectedToReceive,
            "progress": progress.fractionCompleted,
            "status": "downloading"
          ])
        }
      }
      
      // Store task and observer
      self.activeDownloads[modelName] = task
      self.downloadObservers[modelName] = observation
      
      // Start download
      task.resume()
  }

  private func cancelDownload(modelName: String, promise: Promise) {
    if let task = activeDownloads[modelName] {
      task.cancel()
      activeDownloads.removeValue(forKey: modelName)
      downloadObservers.removeValue(forKey: modelName)?.invalidate()
      sendEvent("downloadProgress", ["modelName": modelName, "status": "cancelled"])
      promise.resolve(true)
    } else {
      promise.resolve(false)
    }
  }

  private func createModelFromDownloaded(modelName: String, maxTokens: Int?, topK: Int?, temperature: Double?, randomSeed: Int?, promise: Promise) {
    let modelURL = getModelURL(modelName: modelName)
    if !FileManager.default.fileExists(atPath: modelURL.path) {
      promise.reject("ERR_MODEL_NOT_FOUND", "Model \(modelName) is not downloaded")
      return
    }
    do {
      let handle = try self.createModelInternal(
        modelPath: modelURL.path,
        maxTokens: maxTokens ?? 1024,
        topK: topK ?? 40,
        temperature: temperature ?? 0.7,
        randomSeed: randomSeed ?? 42
      )
      promise.resolve(handle)
    } catch {
      promise.reject("ERR_CREATE_MODEL", "Failed to create model: \(error.localizedDescription)")
    }
  }

  private func createModelInternal(modelPath: String, maxTokens: Int, topK: Int, temperature: Double, randomSeed: Int) throws -> Int {
    let modelHandle = nextHandle
    nextHandle += 1
    let model = try LlmInferenceModel(
      modelPath: modelPath,
      maxTokens: maxTokens,
      topK: topK,
      temperature: Float(temperature),
      randomSeed: randomSeed,
      eventEmitter: { [weak self] eventName, params in self?.sendEvent(eventName, params) },
      modelHandle: modelHandle
    )
    modelMap[modelHandle] = model
    return modelHandle
  }
}
